src/coreclr/src/vm/dllimportcallback.cpp

```
@@ -626,32 +627,93 @@ VOID UMThunkMarshInfo::SetUpForUnmanagedCallersOnly()
+    namedArgs[1].Init("EntryPoint", SERIALIZATION_TYPE_STRING, caEntryPoint);
+
+    InlineFactory<SArray<CaValue>, 4> caValueArrayFactory;
+    DomainAssembly* domainAssembly = pMD->GetLoaderModule()->GetDomainAssembly();
+    IfFailThrow(Attribute::ParseAttributeArgumentValues(
+        pData,
+        cData,
+        &caValueArrayFactory,
+        NULL,
+        0,
+        namedArgs,
+        lengthof(namedArgs),
+        domainAssembly));
+
+    // If the value isn't defined, then return without setting anything.
+    if (namedArgs[0].val.type.tag == SERIALIZATION_TYPE_UNDEFINED)
+        return;
+
+    CorPinvokeMap callConvLocal = (CorPinvokeMap)0;
+    if (nativeCallableInternalData)
+    {
+        callConvLocal = (CorPinvokeMap)(namedArgs[0].val.u4 << 8);
+    }
+    else
+    {
+        // Set WinAPI as the default
+        callConvLocal = CorPinvokeMap::pmCallConvWinapi;
+
+        CaValue* arrayOfTypes = &namedArgs[0].val;
+        for (ULONG i = 0; i < arrayOfTypes->arr.length; i++)
+        {
+            CaValue& typeNameValue = arrayOfTypes->arr[i];
+
+            StackSString typeFQName(SString::Utf8, typeNameValue.str.pStr, typeNameValue.str.cbStr);
+            if (typeFQName.BeginsWith(W("System.Runtime.CompilerServices.CallConvCdecl")))
+            {
+                callConvLocal = CorPinvokeMap::pmCallConvCdecl;
+            }
+            else if (typeFQName.BeginsWith(W("System.Runtime.CompilerServices.CallConvStdcall")))
+            {
+                callConvLocal = CorPinvokeMap::pmCallConvStdcall;
+            }
+            else if (typeFQName.BeginsWith(W("System.Runtime.CompilerServices.CallConvFastcall")))
+            {
+                callConvLocal = CorPinvokeMap::pmCallConvFastcall;
+            }
+            else if (typeFQName.BeginsWith(W("System.Runtime.CompilerServices.CallConvThiscall")))
+            {
+                callConvLocal = CorPinvokeMap::pmCallConvThiscall;
+            }
```

Should we throw for unknown callconv?

---

Definitely an option. I was deferring to the compiler for the most common cases, but being proactive here is probably okay. Is this a `NotSupportedException` or an `InvalidProgramException`?

---

Went with an existing approach about calling conventions.

---

The other places in the VM throw `COMPlusThrow(kTypeLoadException, ...` for invalid calling convention in metadata, e.g. https://github.com/dotnet/runtime/blob/97e553f5c3244b7df65ebfb7de3bb712081228cc/src/coreclr/src/vm/dllimport.cpp#L4138

---

Missed this comment. Will use this one.